airbnb_project/listings/harvester_app/harvester/spiders/listings_spider.py

```diff
@@ -73,117 +75,238 @@ class ListingsSpider(scrapy.Spider):
     # Instance of AirBnbURLBuilder to generate URLs for different search parameters
     URL_BUILDER = AirBnbURLBuilder()
 
-    def start_requests(self):
+    def start_requests(self) -> List[scrapy.FormRequest]:
+        """
+        Generate initial requests for scraping Airbnb listings using coordinates from a JSON file.
+
+        This method orchestrates the process of loading already scraped listings,
+        loading coordinates, and generating requests based on those coordinates.
+
+        Returns:
+            List[scrapy.FormRequest]: A list of FormRequest objects for initial scraping.
+        """
+        self._load_already_scraped_listings()
+        coordinates = self._load_coordinates()
```

Comment:
this is a temporary solution. Once the database is in, the database should/maybe handle the duplicate issue.

Reply:
how do you mean cache? in what way

Reply:
i was thinking you can set a flag to avoid re-reading from json file every run
for example, in `_load_coordinates` method:
```python
if hasattr(self, '_cached_coordinates'):
        return self._cached_coordinates  # Return the cached data

    try:
        current_directory = os.path.dirname(os.path.abspath(__file__))
        coordinates_file_path = os.path.join(current_directory, 'coordinates.json')
        with open(coordinates_file_path, 'r') as file:
            self._cached_coordinates = json.load(file)
    except FileNotFoundError:
        print("Coordinates file not found")
        self._cached_coordinates = []
    except json.JSONDecodeError:
        print("Error decoding JSON in coordinates file")
        self._cached_coordinates = []
    except Exception:
        print("Unknown error occurred while reading the coordinates file")
        self._cached_coordinates = []

    return self._cached_coordinates
```
Does not need to be in this pr, but i guess keep this as suggestion. Let me know

Reply:
this code is only run at the start of the program therefore the flag will always be false.

Reply:
Sorry I incorrectly indented after first if block